File: Olympiad/COCI/official/2007/contest6_solutions/kamen.pas

(*
  Croatian Open Competition in Informatics 2006/2007
  Contest 6 - Task KAMEN
  Programming language Pascal
*)

program kamen;

type
   putanja = record
      stupac : array[1..30000] of integer;
      r : longint;
   end;

var
   RR, SS, N : longint;
   a : array[1..30000, 1..32] of char;
   P : array[1..30] of putanja;

procedure ubaci( var x : putanja );
begin
   a[x.r-1, x.stupac[x.r-1]] := 'O';
end;

procedure sredi( var x : putanja );
var
   s : longint;
begin
   while true do begin
      s := x.stupac[x.r-1];

      if (x.r > 1) and (a[x.r-1,s] <> '.') then begin dec( x.r ); continue; end;

      if x.r > RR then break;
      if a[x.r,s] = 'X' then break;

      if a[x.r,s] = '.' then begin
         x.stupac[x.r] := s;
         inc( x.r );
      end else begin
         if (s > 1) and (a[x.r,s-1] = '.') and (a[x.r-1,s-1] = '.') then begin
            x.stupac[x.r] := s-1;
            inc( x.r );
         end else if (s < SS) and (a[x.r,s+1] = '.') and (a[x.r-1,s+1] = '.') then begin
            x.stupac[x.r] := s+1;
            inc( x.r );
         end else break;
      end;
   end;
end;

var
   i, j : longint;

begin
   readln( RR, SS );
   for i := 1 to RR do readln( a[i] );

   for i := 1 to SS do begin
      P[i].stupac[1] := i;
      P[i].r := 2;
      sredi( P[i] );
   end;

   readln( N );
   for i := 1 to N do begin
      readln( j );
      ubaci( P[j] );
      for j := 1 to SS do sredi( P[j] );
   end;

   for i := 1 to RR do begin
      for j := 1 to SS do
         write( a[i,j] );
      writeln;
   end;
end.

File: Olympiad/COCI/official/2007/contest6_solutions/kamen.cpp

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 6 - Task KAMEN
  Programming language C++
*/

#include <cstdio>

using namespace std;

int R, S, N;
char a[30000][32];

struct putanja {
   char stupac[30000];
   int r;

   void ubaci() {
      a[r-1][stupac[r-1]] = 'O';
   }

   void sredi() {
      for( ;; ) {
         int s = stupac[r-1];

         if( r > 1 && a[r-1][s] != '.' ) { --r; continue; }

         if( r == R ) break;
         if( a[r][s] == 'X' ) break;
         if( a[r][s] == '.' ) {
            stupac[r++] = s;
         } else {
            if( s > 0 && a[r][s-1] == '.' && a[r-1][s-1] == '.' ) {
               stupac[r++] = s-1;
            } else if( s+1 < S && a[r][s+1] == '.' && a[r-1][s+1] == '.' ) {
               stupac[r++] = s+1;
            } else {
               break;
            }
         }
      }
   }
} P[30];

int main( void ) {
   scanf( "%d%d", &R, &S );
   for( int r = 0; r < R; ++r ) scanf( "%s", a[r] );

   for( int i = 0; i < S; ++i ) {
      P[i].stupac[0] = i;
      P[i].r = 1;
      P[i].sredi();
   }

   scanf( "%d", &N );
   for( int i = 0; i < N; ++i ) {
      int s;
      scanf( "%d", &s ); --s;
      P[s].ubaci();
      for( int i = 0; i < S; ++i ) P[i].sredi();
   }

   for( int r = 0; r < R; ++r ) printf( "%s\n", a[r] );

   return 0;
}
